### PR TITLE
Add 'object-literal-contextual-type' rule

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update

#### Overview of change:

Added the `object-literal-contextual-type` rule, which warns for an object literal with no contextual type (which may have unchecked excess properties).
This branch is based on top of #2544 and will be rebased once that's in.

#### CHANGELOG.md entry:

[new-rule] `object-literal-contextual-type`

### CODE REVIEW

**Comment by andy-hanson** [comment]
Rebased.

File: src/rules/completedDocsRule.ts [hunk 1]
```diff
@@ -109,7 +109,7 @@ export class Rule extends Lint.Rules.TypedRule {
         },
     };
 
-    public static ARGUMENT_DESCRIPTOR_BLOCK = {
+    public static ARGUMENT_DESCRIPTOR_BLOCK: {} = {
```
Comment (giladgray) [inline src/rules/completedDocsRule.ts]: this isn't even correct. it has a much more advanced type than the empty object.

File: src/configs/recommended.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@
  * limitations under the License.
  */
 
+// tslint:disable object-literal-contextual-type
```
Comment (giladgray) [inline src/configs/recommended.ts]: API break!! not down for this.

File: src/rules/objectLiteralContextualTypeRule.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        description: "Requires that every object literal has a contextual type.",
+        rationale: Lint.Utils.dedent`
+            An object literal with no type does not have excess properties checked.
+
+            For example:
+
+                interface I { x: number; }
+                function f(): I {
+                    const res = { x: 0, y: 0 };
+                    return res;
```
Comment (giladgray) [inline src/rules/objectLiteralContextualTypeRule.ts]: this error can be avoided so many ways that don't require a rather aggressive lint rule...

Comment (andy-hanson) [reply src/rules/objectLiteralContextualTypeRule.ts]: How would it be avoided in general? I don't believe typescript's behavior with respect to object literals has changed much in the past year.
Often people will think they're getting a contextual type when they really aren't, e.g. a result from a callback:
```ts
interface Options { foo?: number; bar?: number }
const arr: Options[] = [1, 2, 3].map(n => ({ foo: n, baa: n }));
```
The error is fixed by changing to `(n): Options`, but I wouldn't want a lint rule to require all arrow functions to have explicit return types.
If there are better ways of avoiding this error I'd like to hear about them.

File: src/configs/recommended.ts (hunk 2 above)
Comment (andy-hanson) [reply src/configs/recommended.ts]: This is a comment, how does it break the API?

File: src/rules/completedDocsRule.ts (hunk 1 above)
Comment (andy-hanson) [reply src/rules/completedDocsRule.ts]: `{}` is a correct type. Anything but `null` or `undefined` or `void` is assignable to `{}`.

**Pull request closed without merging**